synapse/storage/databases/main/keys.py

```diff
@@ -205,35 +205,38 @@ async def get_server_keys_json_for_remote(
 
         If we have multiple entries for a given key ID, returns the most recent.
         """
-        rows = await self.db_pool.simple_select_many_batch(
-            table="server_keys_json",
-            column="key_id",
-            iterable=key_ids,
-            keyvalues={"server_name": server_name},
-            retcols=(
-                "key_id",
-                "from_server",
-                "ts_added_ms",
-                "ts_valid_until_ms",
-                "key_json",
+        rows = cast(
+            List[Tuple[str, str, int, int, Union[bytes, memoryview]]],
+            await self.db_pool.simple_select_many_batch(
+                table="server_keys_json",
+                column="key_id",
+                iterable=key_ids,
+                keyvalues={"server_name": server_name},
+                retcols=(
+                    "key_id",
+                    "from_server",
+                    "ts_added_ms",
+                    "ts_valid_until_ms",
+                    "key_json",
+                ),
+                desc="get_server_keys_json_for_remote",
             ),
-            desc="get_server_keys_json_for_remote",
         )
 
         if not rows:
             return {}
 
-        # We sort the rows so that the most recently added entry is picked up.
-        rows.sort(key=lambda r: r["ts_added_ms"])
+        # We sort the rows by ts_added_ms so that the most recently added entry is picked up.
+        rows.sort(key=lambda r: r[2])
```

Comment:
I find this a bit less clear, but I think the rest of the improvements outweigh it.

I'm not really sure why we even sort this if we put it into a mapping anyway though, this seems like useless work.

Reply:
> Dictionaries preserve insertion order these days... but agreed that it seems odd.

Right, but we usually don't depend on that?

Reply:
Agreed.

Reply:
(or if we do, we explicitly use an OrderedDict, I think?)

Reply:
I guess I'll just leave it for now then to not break anything.

Reply:
It was added in #16123, it exists because the key ID might exist multiple times.